cluster-autoscaler/utils/drain/drain.go

```diff
@@ -0,0 +1,192 @@
+				// Assume a reason for an error is because the RC is either
+				// gone/missing or that the rc has too few replicas configured.
+				// TODO: replace the minReplica check with pod disruption budget.
+				if err == nil && rc != nil {
+					if rc.Spec.Replicas < minReplica {
+						return []*api.Pod{}, fmt.Errorf("replication controller for %s/%s has too few replicas spec: %d min: %d",
+							pod.Namespace, pod.Name, rc.Spec.Replicas, minReplica)
+					} else {
+						replicated = true
+					}
+				} else {
+					return []*api.Pod{}, fmt.Errorf("replication controller for %s/%s is not available, err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				replicated = true
+			}
+		} else if refKind == "DaemonSet" {
+			if checkReferences {
+				ds, err := client.DaemonSets(sr.Reference.Namespace).Get(sr.Reference.Name)
+
+				// Assume the only reason for an error is because the DaemonSet is
+				// gone/missing, not for any other cause.  TODO(mml): something more
+				// sophisticated than this
+				if err == nil && ds != nil {
+					// Otherwise, treat daemonset-managed pods as unmanaged since
+					// DaemonSet Controller currently ignores the unschedulable bit.
+					// FIXME(mml): Add link to the issue concerning a proper way to drain
+					// daemonset pods, probably using taints.
+					daemonsetPod = true
+				} else {
+					return []*api.Pod{}, fmt.Errorf("deamonset for %s/%s is not present, err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				daemonsetPod = true
+			}
+		} else if refKind == "Job" {
+			if checkReferences {
+				job, err := client.ExtensionsClient.Jobs(sr.Reference.Namespace).Get(sr.Reference.Name)
+
+				// Assume the only reason for an error is because the Job is
+				// gone/missing, not for any other cause.  TODO(mml): something more
+				// sophisticated than this
+				if err == nil && job != nil {
+					replicated = true
+				} else {
+					return []*api.Pod{}, fmt.Errorf("job for %s/%s is not available: err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				replicated = true
+			}
+		} else if refKind == "ReplicaSet" {
+			if checkReferences {
+				rs, err := client.ExtensionsClient.ReplicaSets(sr.Reference.Namespace).Get(sr.Reference.Name)
+
+				// Assume the only reason for an error is because the RS is
+				// gone/missing, not for any other cause.  TODO(mml): something more
+				// sophisticated than this
+				if err == nil && rs != nil {
+					if rs.Spec.Replicas < minReplica {
+						return []*api.Pod{}, fmt.Errorf("replication controller for %s/%s has too few replicas spec: %d min: %d",
+							pod.Namespace, pod.Name, rs.Spec.Replicas, minReplica)
+					} else {
+						replicated = true
+					}
+				} else {
+					return []*api.Pod{}, fmt.Errorf("replication controller for %s/%s is not available, err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				replicated = true
+			}
+		}
+		if !daemonsetPod && !replicated {
```

Comment:
You are checking in each case whether the pod is not created by daemon controller. How about just checking it once and continue?